File: zivid_samples/scripts/sample_capture.py
#!/usr/bin/env python

import sys

from rcl_interfaces.srv import SetParameters
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.parameter import Parameter
from sensor_msgs.msg import PointCloud2
from std_srvs.srv import Trigger


class Sample(Node):

    def __init__(self):
        super().__init__('sample_capture_py')

        self.capture_service = self.create_client(Trigger, 'capture')
        while not self.capture_service.wait_for_service(timeout_sec=3.0):
            self.get_logger().info('capture service not available, waiting again...')

        self._set_settings()

        self.subscription = self.create_subscription(
            PointCloud2, 'points/xyzrgba', self.on_points, 10
        )

    def _set_settings(self):
        self.get_logger().info('Setting parameter `settings_yaml`')
        settings_parameter = Parameter(
            'settings_yaml',
            Parameter.Type.STRING,
            """
__version__:
  serializer: 1
  data: 22
Settings:
  Acquisitions:
    - Acquisition:
        Aperture: 5.66
        ExposureTime: 8333
  Processing:
    Filters:
      Outlier:
        Removal:
          Enabled: yes
          Threshold: 5
""",
        ).to_parameter_msg()

        param_client = self.create_client(SetParameters, 'zivid_camera/set_parameters')
        while not param_client.wait_for_service(timeout_sec=3):
            self.get_logger().info('Parameter service not available, waiting again...')

        future = param_client.call_async(
            SetParameters.Request(parameters=[settings_parameter])
        )
        rclpy.spin_until_future_complete(self, future, timeout_sec=30)
        if not future.result():
            raise RuntimeError('Failed to set parameters')

    def capture(self):
        self.get_logger().info('Calling capture service')
        return self.capture_service.call_async(Trigger.Request())

    def on_points(self, msg):
        self.get_logger().info(
            f'Received point cloud of size {msg.width} x {msg.height}'
        )


def main(args=None):
    rclpy.init(args=args)

    try:
        sample = Sample()

        sample.get_logger().info('Spinning node.. Press Ctrl+C to abort.')

        while rclpy.ok():
            future = sample.capture()
            rclpy.spin_until_future_complete(sample, future)
            sample.get_logger().info('Capture complete')

    except KeyboardInterrupt:
        pass
    except ExternalShutdownException:
        sys.exit(1)


if __name__ == '__main__':
    main()

File: zivid_samples/src/sample_capture.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <std_srvs/srv/trigger.hpp>
#include <stdexcept>

/*
 * This sample shows how to set the settings_file_path parameter of the zivid node, subscribe for
 * the points/xyzrgba topic, and invoke the capture service. When a point cloud is received, a new
 * capture is triggered.
 */

class CaptureHandler {
public:
  CaptureHandler(rclcpp::Node::SharedPtr node)
  : node_(node),
    capture_client_(node_->create_client<std_srvs::srv::Trigger>("capture")) 
  {
    subscription_ = node_->create_subscription<sensor_msgs::msg::PointCloud2>(
      "points/xyzrgba", 10,
      std::bind(&CaptureHandler::on_point_cloud, this, std::placeholders::_1));

    wait_for_service();
  }

  void wait_for_service()
  {
    while (!capture_client_->wait_for_service(std::chrono::seconds(3))) {
      if (!rclcpp::ok()) {
        fatal_error("Interrupted while waiting for capture service.");
      }
      RCLCPP_INFO(node_->get_logger(), "Waiting for capture service...");
    }
    RCLCPP_INFO(node_->get_logger(), "Capture service is available.");
  }

  void trigger_capture() {
    RCLCPP_INFO(node_->get_logger(), "Triggering capture");
    auto request = std::make_shared<std_srvs::srv::Trigger::Request>();
    capture_client_->async_send_request(request);
  }

  void on_point_cloud(const sensor_msgs::msg::PointCloud2::ConstSharedPtr & msg) {
    RCLCPP_INFO(node_->get_logger(), "Got point cloud: %d x %d", msg->width, msg->height);
    trigger_capture();
  }

private:
  void fatal_error(const std::string & message) {
    RCLCPP_ERROR_STREAM(node_->get_logger(), message);
    throw std::runtime_error(message);
  }

  rclcpp::Node::SharedPtr node_;
  rclcpp::Client<std_srvs::srv::Trigger>::SharedPtr capture_client_;
  rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr subscription_;
};

void set_settings(const std::shared_ptr<rclcpp::Node> & node)
{
  const std::string settings_yml = R"(
__version__:
  serializer: 1
  data: 22
Settings:
  Acquisitions:
    - Acquisition:
        Aperture: 5.66
        ExposureTime: 8333
  Processing:
    Filters:
      Outlier:
        Removal:
          Enabled: yes
          Threshold: 5
)";
  auto param_client = std::make_shared<rclcpp::AsyncParametersClient>(node, "zivid_camera");

  while (!param_client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      throw std::runtime_error("Interrupted while waiting for parameter service.");
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for parameter service...");
  }

  auto result = param_client->set_parameters({rclcpp::Parameter("settings_yaml", settings_yml)});
  if (rclcpp::spin_until_future_complete(node, result, std::chrono::seconds(120)) !=
      rclcpp::FutureReturnCode::SUCCESS) {
    throw std::runtime_error("Failed to set `settings_yaml` parameter");
  }
}

void set_srgb(const std::shared_ptr<rclcpp::Node> & node)
{
  auto param_client = std::make_shared<rclcpp::AsyncParametersClient>(node, "zivid_camera");

  while (!param_client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      throw std::runtime_error("Interrupted while waiting for parameter service.");
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for parameter service...");
  }

  auto result = param_client->set_parameters({rclcpp::Parameter("color_space", "srgb")});
  if (rclcpp::spin_until_future_complete(node, result, std::chrono::seconds(30)) !=
      rclcpp::FutureReturnCode::SUCCESS) {
    throw std::runtime_error("Failed to set `color_space` parameter");
  }
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("sample_capture");

  RCLCPP_INFO(node->get_logger(), "Started the sample_capture node");

  set_settings(node);
  set_srgb(node);

  auto handler = std::make_shared<CaptureHandler>(node);

  handler->trigger_capture();

  RCLCPP_INFO(node->get_logger(), "Spinning node.. Press Ctrl+C to abort.");
  rclcpp::spin(node);
  rclcpp::shutdown();

  return EXIT_SUCCESS;
}
